bufit.go

```diff
@@ -145,6 +145,18 @@ func (b *Buffer) Close() error {
 	return nil
 }
 
+// Len returns the #  of bytes buffered for Readers
+func (b *Buffer) Len() int {
+	return b.buf.Len()
+}
+
+// Discard drops the next n buffered bytes. It returns the actual number of
+// bytes dropped and may return io.EOF if all remaining bytes have been
+// discarded.
+func (b *Buffer) Discard(n int) (int, error) {
+	return b.buf.Discard(n)
```

Comment:
In this case, the internal Writer (buf) might Discard bytes that are being pointed to by readers.
If you look at the implementation of "shift" you'll see that buf.Discard is only called on bytes which have been read by the "Peek" (reader that has the earliest offset in the shared buffer). This guarantees that no reader can be pointing to those bytes.

My laptop is about to run out of battery but once I get home I'll try to share some ideas on how we can still get some control over the buffer size (back pressure might be our only other option here, dropping data is hard since those bytes might be pointed to by a concurrent reader).